### PR TITLE
The Great `refiner/refiner.fun` Schism

### PR DESCRIPTION
**Pull request by jozefg**
This PR breaks out `refiner.fun` into a number of separate functors. Instead of having the rules inline, each of the connectives (and one file for connective-independent tactics) gets it's own functor. This functor takes in `RULES_UTIL` which contains all the helper functions/other goodies found originally in `refiner.fun`. The actual `refiner.fun` now just reads as a list of "Load the rules for this connective, inject them into our scope".

Workflow change: instead of modifying `refiner.fun` directly one should modify the appropriate functor for the connective. There would be two signatures to change `refiner.sig` and the signature for the particular connective. This duplication is avoidable (we could have `refiner.sig` just be includes) I'm on the fence about whether this is the correct move (Jon: do you think we should do this? If so I can make this change easily enough).

If you need to create a new connective, you create a new functor. Follow the template they all start with and it's easy enough.

Note: `BHyp` is a little strange since it depends on other tactics, everything else was relatively easy though.

### CODE REVIEW

File: src/refiner/refiner.fun [hunk 1]
```diff
@@ -36,1640 +36,79 @@ struct
...
-
-  infix $ \
-  infix 8 $$ //
-  infixr 8 \\
-
   exception Refine
 
   structure Rules =
   struct
-    open Goal Sequent
```
Comment (jozefg) [inline src/refiner/refiner.fun:62]: This part of the diff makes me very happy.

**Comment by jonsterling** [comment]
This is glorious!!

BTW, my preference would be to use includes rather than writing the signatures twice. However, is it possible to do that in this context? Would you get "duplicate type" errors for all the type members of the signatures (e.g. `type tactic`)?

Another option would be to have `REFINER` be hierarchical, like

``` sml
signature REFINER = 
sig
  ...
  structure IsectRules : ISECT_RULES
  ...
```

or whatever. In fact, this might be a better option in the long run, since it would give us the ability to have the individual signatures have functions with shorter names, like `IsectRules.Intro` instead of `IsectRules.IsectIntro`, etc. I don't feel too strongly either way though, so feel free to choose whatever approach you prefer.

**Comment by jonsterling** [comment]
One change that might be worthwhile is to move things like `bhyp`, which are really proper tactics and not rules, from the proper rules. Feel free either to do it in this PR, or to put it off for later; I wouldn't hold up the merge of this excellent changeset for that.

**Comment by jozefg** [comment]
@jonsterling I think you can get around the duplicated type errors by aggressively `where type`-ing. I'm kind of inclined to make this just `include` for now because this makes the change completely unnoticeable outside of `refiner.*`, it's easy enough to move to nested modules anyways.

As far as `BHyp`, perhaps I'll hold up on that until next time :)

Just moving the comments from `refiner.sig` to the respective signatures and then we'll be all set :)

**Comment by jonsterling** [comment]
sounds great! thanks again!

**Comment by jozefg** [comment]
@jonsterling Whoops, so it turns out I was mistaken about those `includes`, redundant type declarations abound. To solve this I just went with the "lots of embedded structures approach". I also took the liberty of removing all the now redundant prefixes `ApproxRules.ApproxRefl` is now just `ApproxRules.Refl`.

I think it's about ready to merge!

**Comment by jonsterling** [comment]
A truly herculean effort! Many thanks!

**Pull request merged**